keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2861 @@
+    - Trigger `FailurePolicy` on all admission validations
+    - Add a detailed error in the status of the `ValidatingAdmissionPolicy`.
+
+A partial `spec.match` schema (subset of the full schema) is okay so long as
+only optional fields are omitted. But any unrecognized field in the `spec.match`
+would not be allowed. 
+
+Proposed annotation: 
+
+> Example: `admission.kubernetes.io/is-policy-configuration-definition: "true"`
+> 
+> Used on: CustomResourceDefinition
+> 
+> What a CustomResourceDefinition has the annotation set to "true", the
+> OpenAPIv3 schema of all versions of this resource is modified and then
+> kept-in-sync by a controller to always contain the expected schema fields of
+> admission policy configuration resources.
+
+xref: https://kubernetes.io/docs/reference/labels-annotations-taints/
+
+Pros:
+
+- A single resource is used to configure both the match criteria and
+  configuration params of a policy.
+- Provides a clear marker on every CRD that is used for policy configuration
+  that it is intended for this purpose, in a machine readable way.
+
+Cons:
+
+- API server must check for a wide range of error conditions and define how
+  exactly it handles each of them.
+- If the `spec.match` schema is incorrectly defined, CRD author might not
+  realize it since they need to check the status of the corresponding
+  `ValidatingAdmissionPolicy` for any errors.
+
+###### Alternative: `PolicyConfiguration` kind with config reference
+
+Require two resources:
+
+- A `PolicyConfiguration` resources contains the match criteria and a reference
+  to a configuration resource.
+- A configuration resource that contains only configuration (no match criteria).
+
+Example:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: PolicyConfiguration
+metadata:
+  name: "xyzlimit-scale.example.com"
+  labels:
+    policyConfig: xyzlimit
+spec:
+  match:
+    resourceRules:
+      - apiGroups: ["*"]
+        apiVersions: ["*"]
+        resources: ["*/scale"]
+  config:
+    name: xyzlimit-scale-config.example.com
+```
+
+```yaml
+apiVersion: rules.example.com/v1
+kind: XyzLimit
+metadata:
+  name: "xyzlimit-config.example.com"
+spec:
+  allowed: ["a", "b", "c"]
+  banned: ["x", "y", "z"]
+  xyz:
+    fuzzFactor: 0.8
+    reticulate: true
+```
+
+Note that for policy configurations that require no parameterization, only the
+`PolicyConfiguration` is needed.
+
+The `metadata.labels.policyConfig` label on `PolicyConfiguration` is needed so
+that each `ValidatingAdmissionPolicy` use a label selector to reference the
+policy configurations that configure it.
+
+Pros:
+
+- Matching criteria is fully defined and validated in a builtin type.
+- The configuration data that the policy definition is treated as an opaque
+  payload from an apiserver perspective, it passes it to the policy definition
+  so that CEL expressions can access it, but there is no code in the apiserver
+  that cares about the structure of the configuration CRD.
```

Comment:
? You can load the CRD's schema and type-check when you do the other typechecking?

Reply:
(Also if this is true it should be a con, but I don't see why it would be true)

Reply:
I think this alternative makes fully type checking CEL expressions the easiest.

Reply:
I agree about type checking. I think I failed to make a point here. I was trying to note that unlike the "duck typing" alternative, the API server is not required to have any expectations about the structure of the CRD for this design, and that makes things simpler.

I'll rephrase and add a note about type checking.